source/util/timer.h

```diff
@@ -0,0 +1,332 @@
+// kClockGettimeFailed means it failed in calling clock_gettime().
+enum UsageStatus { kGetrusageFailed, kClockGettimeFailed, kSucceeded };
+
+// Timer measures the resource utilization for a range of code. The resource
+// utilization consists of CPU time (i.e., process time), WALL time (elapsed
+// time), USR time, SYS time, RSS, and the number of page faults. RSS and the
+// number of page faults are measured only when |measure_mem_usage| given to the
+// constructor is true. This class should be used as the following example:
+//
+//   spvutils::Timer timer(std::cout);
+//   timer.Start();       // <-- set |usage_before_|, |wall_before_|,
+//                               and |cpu_before_|
+//
+//   /* ... lines of code that we want to know its resource usage ... */
+//
+//   timer.Stop();        // <-- set |cpu_after_|, |wall_after_|, and
+//                               |usage_after_|
+//   timer.Report(tag);   // <-- print tag and the resource utilization to
+//                               std::cout.
+class Timer {
+ public:
+  Timer(std::ostream* out, bool measure_mem_usage = false)
+      : report_stream_(out),
+        usage_status_(kSucceeded),
+        measure_mem_usage_(measure_mem_usage) {}
+
+  // Sets |usage_before_|, |wall_before_|, and |cpu_before_| as results of
+  // getrusage(), clock_gettime() for the wall time, and clock_gettime() for the
+  // CPU time respectively. Note that this method erases all previous state of
+  // |usage_before_|, |wall_before_|, |cpu_before_|.
+  virtual void Start();
+
+  // Sets |cpu_after_|, |wall_after_|, and |usage_after_| as results of
+  // clock_gettime() for the wall time, and clock_gettime() for the CPU time,
+  // getrusage() respectively. Note that this method erases all previous state
+  // of |cpu_after_|, |wall_after_|, |usage_after_|.
+  virtual void Stop();
+
+  // If |report_stream_| is NULL, it does nothing. Otherwise, it prints the
+  // resource utilization (i.e., CPU/WALL/USR/SYS time, RSS) from the time of
+  // calling Timer::Start() to Timer::Stop().
+  void Report(const char* tag);
+
+  // Returns the measured CPU Time (i.e., process time) for a range of code
+  // execution.
+  virtual double CPUTime() { return TimeDifference(cpu_before_, cpu_after_); }
+
+  // Returns the measured Wall Time (i.e., elapsed time) for a range of code
+  // execution.
+  virtual double WallTime() {
+    return TimeDifference(wall_before_, wall_after_);
+  }
+
+  // Returns the measured USR Time for a range of code execution.
+  virtual double UserTime() {
+    return TimeDifference(usage_before_.ru_utime, usage_after_.ru_utime);
+  }
+
+  // Returns the measured SYS Time for a range of code execution.
+  virtual double SystemTime() {
+    return TimeDifference(usage_before_.ru_stime, usage_after_.ru_stime);
+  }
+
+  // Returns the measured RSS for a range of code execution.
+  virtual long RSS() const {
+    return usage_after_.ru_maxrss - usage_before_.ru_maxrss;
```

Comment:
Please state that this returns the RSS delta since the pass started.  Label the column as 'RSS delta' too.  Otherwise, it can be confusing.

Reply:
Done